### ISSUE TITLE
PSDReader incorrectly uses grayscale ICCProfile with a grayscale background.

### ISSUE THREAD

**Issue by Eurybiadan** [haraldk/TwelveMonkeys#244]
Hello again,

I've found a (rather specific bug) that I see in my application of the PSDReader. I'm pretty sure know exactly what the problem is, but I'm not sure how you want to approach fixing it (or I'd just submit a pull request...).

As the title suggests, the reader does not correctly use the ICC Profile of Grayscale PSD files that have a grayscale background. When a background is enabled (visible) in the attached psd (renamed as a png, of course), the intensities are washed out, consistent with a basic linear grayscale colorspace, instead of the more typical Gamma 2.2 or Dot Gain 20%.

This appears to be because the number of channels changes when there is a background. When the background is invisible (transparent), the file contains 2 channels. When the background is visible, the file contains 1 channel. This causes a variety of issues that causes the simple "ImageIO.read" to fail (i.e. load the wrong bands).

I attempted to get around this bug by creating my own ImageReadParam, but it appears that even if the data gets loaded correctly, the destination color model I choose with my ImageReadParam is not being applied (PSDImageReader:line 496) for some reason, which still causes the layers to appear washed out.

As far as fixes go, I see two options:

1) Honor every aspect of the ImageReadParam that is submitted, down to the color model, and don't convert after the fact.
2) Change lines 174 and 180 of PSDImageReader to be consistent with the other channel/bit levels, ex: `return ImageTypeSpecifiers.createBanded(cs, new int[] {0}, new int[] {0}, DataBuffer.TYPE_BYTE, false, false);`

I have workarounds like forcing the grayscale image into a more "typical" curve via LUTs (ala gamma 2.2 or Dot Gain 20%), but it would be nice if I didn't have to detect if the user has a visible background (and all the handling code that entails).

Respectfully,

RFC

![itty_bitty](https://cloud.githubusercontent.com/assets/5322065/15309015/e07a6b8a-1bb0-11e6-94c6-e079de4efe9d.png)

**Comment by haraldk** [comment]
Hi Robert,

Thanks for the detailed report!

I think that ignoring the ICC profile for grayscale for single channel images is a bug/omission from my side, so we should definitively do option 2). For your test image, this seems to create colors consistent with what I see in Preview on OS X. 

Supplying a different destination color space in the `ImageReadParam` probably won't help, as the source color space is already bogus (i.e., the pixel data will be *converted* from the incorrect color model, to the correct one, but that means the colors will still *look* the same as before the conversion. There's no way to set te source color space using an `ImageReadParam`). However, if we fix 2), I think this will also work as expected.

If you have the time, feel free to create a pull request, otherwise I'll be able fix this easily thanks to your excellent analysis. 

I'm just curious/worried about this line:

> This causes a variety of issues that causes the simple "ImageIO.read" to fail (i.e. load the wrong bands).

Is this a separate bug? Are you saying that using `ImageIO.read(...)` you'll get a different looking image, or is loading different/incorrect bands? Do you have a test program/file to show this? Or am I just missing something? 


Thanks,

-- 
Harald K

**Comment by Eurybiadan** [comment]
Hi Harald,

I'm still working on the underlying cause of the incorrect band detection when using just "ImageIO.read". I'm pretty sure it is related to the "number of channels" detection (similar to the colorspace bug), but I've yet to lock down its exact source. I'll take a closer look today and let you know.

**Comment by Eurybiadan** [comment]
It appears to happening on PSDImageReader: line 1119.

The PSDHeader reads in the number of channels as "1", then the logic there sort of "trips" on itself, and instead of reading in bands 0 and 1, reads in bands 2 and 1.

**Comment by Eurybiadan** [comment]
The index creation section is this:

```
for (int i = 0, indicesLength = indices.length; i < indicesLength; i++) {
     indices[i] = indicesLength - i;
}
```
And when the number of channels is 2 but was detected as 1, it fills in the index array as follows:
```
indicies[0] = 2;
indicies[1] = 1;
```

Changing this to:
```
for (int i = 0; i < indices.length; i++) {
     indices[i] = i;
}
```
Fixes the bug in this case, but I don't want to cause a regression- do you perchance recall why it was constructed in that way?

**Comment by haraldk** [comment]
Hmmm..

Can you please create a separate bug for this, and also supply a failing test case? Preferably JUnit test, but a small class with a main method works too. I also need a sample file to go with the code.

Because I still don't understand how `ImageIO.read(...)` will ever get to execute those lines of code... It should only read the composite image, not start decoding layers.


Thanks,

-- 
Harald K

**Comment by Eurybiadan** [comment]
I'm on it, will try and get one for you this weekend.

Regards,

RFC

**Comment by haraldk** [comment]
Fixed via #247
